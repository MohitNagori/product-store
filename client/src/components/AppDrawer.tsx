import React from 'react';
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { withStyles } from '@material-ui/styles';
import PropTypes from 'prop-types';
import { IconButton, ListItemText, List, Drawer, Divider, ListItemIcon, ListItem }  from '@material-ui/core';
import { ChevronLeftIcon, ChevronRightIcon, MarkunreadMailboxIcon, CategoryIcon } from '../utils/icons'
import { CONSTANTS } from '../utils/constants';
import { IApp } from '../interfaces';
import { appActions as actions } from '../redux/actions';

// Defining Own Props to avoid unwanted props
interface IOwnProps {
  classes?: any;
  theme?: any;
}
type IProps = IOwnProps &
  IApp.StateToProps & 
  IApp.DispatchFromProps;

// Defining Own State to avoid unwanted props
interface IState {}

// Mapping Store data with props 
const mapStateToProps = function(state: any){
  return {
    isDrawerOpen: state.app.isDrawerOpen,
  }
};

// Drawer Component
class AppDrawer extends React.Component<IProps, IState> {
  static propTypes = {
    classes: PropTypes.object.isRequired,
  }

  // Funtion that return icon based on side menu 
  getMenuListIcon = (icon: any) => {
    switch(icon.toLowerCase()) {
      case CONSTANTS.PRODUCTS: 
        return <ListItemIcon><MarkunreadMailboxIcon /></ListItemIcon>;
      case CONSTANTS.CATEGORIES: 
        return <ListItemIcon><CategoryIcon /></ListItemIcon>;
    }
  }

  // Function to render UI
  render() {
    const { classes, theme } = this.props;
    return (
        <Drawer
          className={classes.drawer}
          variant="persistent"
          anchor="left"
          open={this.props.isDrawerOpen}
          classes={{
            paper: classes.drawerPaper,
          }}>
          <div className={classes.drawerHeader}>
            <IconButton onClick={this.props.toggleDrawer}>
              {theme.direction === 'ltr' ? <ChevronLeftIcon /> : <ChevronRightIcon />}
            </IconButton>
          </div>
          <Divider />
          <List>
            {CONSTANTS.MENU_ITEM.map((menuItem, index) => (
              <ListItem key={menuItem.text} className={classes.menuItem}>
                <Link to={menuItem.path} className={classes.link}>
                  { this.getMenuListIcon(menuItem.text) }
                  <ListItemText primary={menuItem.text} />  
                </Link>
              </ListItem>
            ))}
          </List>
          <Divider />
        </Drawer>    
    );
  }
}

// High Order Component to connect with store and action mapping
const AppDrawerWrapper = connect(mapStateToProps, actions)(AppDrawer);

// styles
const styles = {
  drawer: {
    width: CONSTANTS.DRAWER_WIDTH,
    flexShrink: 0,
  },
  drawerPaper: {
    width: CONSTANTS.DRAWER_WIDTH,
  },
  drawerHeader: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  menuItem: {
    padding: 0
  },
  link: {
    display: 'flex',
    textDecoration: 'none !important',
    alignItems: 'center',
    padding: 10,
    width: '100%'
  },
}

/* High Order Component to apply styles as styles can either be applied in HOOK (not in React Component) 
  or with Higher Order Component. We used higher order component as we have used material components.
*/
const AppDrawerWithStyles = withStyles(styles)(AppDrawerWrapper)

export { AppDrawerWithStyles as AppDrawer }
