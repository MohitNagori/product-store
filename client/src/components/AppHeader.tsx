import React from 'react';
import clsx from 'clsx';
import { connect } from "react-redux";
import PropTypes from 'prop-types';
import { Typography, IconButton, Toolbar, AppBar }  from '@material-ui/core';
import { withStyles } from '@material-ui/styles';
import { colors } from '../styles';
import { MenuIcon } from '../utils/icons'
import { CONSTANTS } from '../utils/constants';
import { IApp } from '../interfaces';
import { appActions as actions } from '../redux/actions';

// Defining Own Props to avoid unwanted props
interface IOwnProps {
  classes?: any;
}
type IProps = IOwnProps &
  IApp.StateToProps & 
  IApp.DispatchFromProps;

// Defining Own State to avoid unwanted props
interface IState {
  
}

// Mapping Store data with props 
const mapStateToProps = function(state: any){
  return {
    isDrawerOpen: state.app.isDrawerOpen,
  }
};

// Header Component
class AppHeader extends React.Component<IProps, IState> {
  static propTypes = {
    classes: PropTypes.object.isRequired,
  }
  
  // Function to render UI
  render() {
    const { classes } = this.props;
    return (
        <AppBar
        position="fixed"
        className={clsx({
          [classes.appBarShift]: this.props.isDrawerOpen,
        })}>
        <Toolbar>
          <IconButton
            color="inherit"
            aria-label="open drawer"
            onClick={this.props.toggleDrawer}
            edge="start"
            className={clsx(this.props.isDrawerOpen && classes.hide)}>
            <MenuIcon />
          </IconButton>
          <Typography variant="h6" noWrap>
            Product Blog  
          </Typography>
        </Toolbar>
      </AppBar>
    );
  }
}
// High Order Component to connect with store and action mapping
const AppHeaderWrapper = connect(mapStateToProps, actions)(AppHeader);

// styles
const styles = {
  appBarShift: {
    width: `calc(100% - ${CONSTANTS.DRAWER_WIDTH}px)`,
  },
  hide: {
    display: 'none',
  },
  siteLink: {
    marginLeft: 'auto',
    order: 2,
    textDecoration: 'inherit',
    color: colors.WHITE,
  }
}

/* High Order Component to apply styles as styles can either be applied in HOOK (not in React Component) 
  or with Higher Order Component. We used higher order component as we have used material components.
*/
const AppHeaderWithStyles = withStyles(styles)(AppHeaderWrapper)

export { AppHeaderWithStyles as AppHeader }
