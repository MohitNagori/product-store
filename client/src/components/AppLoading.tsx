import React from 'react';
import clsx from 'clsx';
import { connect } from "react-redux";
import PropTypes from 'prop-types';
import { CircularProgress, Box } from '@material-ui/core';
import { withStyles } from '@material-ui/styles';
import { CONSTANTS } from '../utils/constants';
import { IApp } from '../interfaces';

// Defining Own Props to avoid unwanted props
interface IOwnProps {
    classes?: any;
    loading?: boolean;
}
type IProps = IOwnProps & 
    IApp.StateToProps;

// Defining Own State to avoid unwanted props
interface IState {}

// Mapping Store data with props 
const mapStateToProps = function(state: any){
    return {
      isDrawerOpen: state.app.isDrawerOpen,
    }
};  

// Loading Component
class AppLoading extends React.Component<IProps, IState> {
    static propTypes = {
        classes: PropTypes.object.isRequired,
    }

    // Function to render UI
    render() {
        const { classes } = this.props;
        return (
            <div> 
                { this.props.loading ? 
                    <Box className={clsx(classes.loaderContainer, {
                        [classes.contentShift]: this.props.isDrawerOpen,
                        })} position={'fixed'}>
                        <CircularProgress /> 
                    </Box> : null }
            </div>
        );    
    }
}
// High Order Component to connect with store and action mapping
const AppLoadingWrapper = connect(mapStateToProps, null) (AppLoading);

// styles
const styles = {
    loaderContainer: {
        width: '100%',
        height: 'calc(100vh - 60px)',
        margin: '-20px -25px',
        'background-color': 'rgba(0,0,0,0.4)',
        'z-index': 1,
        display: 'flex',
        'align-items': 'center',
        'justify-content': 'center',
    },
    contentShift: {
        width: `calc(100% - ${CONSTANTS.DRAWER_WIDTH}px)`,
    }
}

/* High Order Component to apply styles as styles can either be applied in HOOK (not in React Component) 
  or with Higher Order Component. We used higher order component as we have used material components.
*/
const AppLoadingWithStyles = withStyles(styles)(AppLoadingWrapper)


export { AppLoadingWithStyles as AppLoading }