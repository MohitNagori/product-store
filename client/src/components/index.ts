// index.ts to assemble all components

export * from './App';
export * from './AppHeader';
export * from './AppDrawer';
export * from './AppMain';
export * from './AppLoading';
