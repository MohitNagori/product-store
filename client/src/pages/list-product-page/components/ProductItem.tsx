import React from 'react';
import { connect } from "react-redux";
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { Card, Avatar, CardHeader, CardContent }  from '@material-ui/core';
import { red, blue } from '@material-ui/core/colors';
import { IProduct } from '../../../interfaces';

// Defining Own Props to avoid unwanted props
interface IOwnProps {
    classes?: any;
    product: IProduct.IProductData;
}
type IProps = IOwnProps;

// Defining Own State to avoid unwanted props
interface IState {

}

// Product Item Page
class ProductItem extends React.Component<IProps, IState> {

    render() {
        const { product, classes } = this.props;
        return (
            <Card className={classes.card}>
                <CardHeader
                    avatar={
                    <Avatar aria-label="recipe" className={classes.avatar}>
                    {product.title[0].toUpperCase()}
                    </Avatar>}
                    title={product.title.split(' ')[0]}
                    subheader={product.subcategory}/>
                <CardContent>
                    <p>Model: {product.title}</p>
                    <p>Price: {product.price}</p>
                    <p>Popularity: {product.popularity}</p>
                </CardContent>
            </Card> 
        );    
    }
}
// High Order Component to connect with store and action mapping
const ProductItemWrapper = connect(null, null) (ProductItem);

const styles = {
    card: {
        maxWidth: '100%',
        margin: 10
    },
    media: {
        height: 0,
        paddingTop: '56.25%',
    },
    link: {
        marginLeft: 'auto',
        order: 2,
        color: blue[500],
        padding: 5
    },
    avatar: {
        backgroundColor: red[500],
    },
}

// Defining proptery to Higher Order Component
ProductItemWrapper.propTypes = {
classes: PropTypes.object.isRequired,
}

/* High Order Component to apply styles as styles can either be applied in HOOK (not in React Component) 
  or with Higher Order Component. We used higher order component as we have used material components.
*/
const ProductItemWithStyles = withStyles(styles)(ProductItemWrapper)

export { ProductItemWithStyles as ProductItem } 