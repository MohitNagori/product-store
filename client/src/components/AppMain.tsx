import React from 'react';
import clsx from 'clsx';
import { connect } from "react-redux";
import { Route } from 'react-router-dom';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/styles';
import { Box } from '@material-ui/core';
import { AppLoading } from './AppLoading';
import { ListProduct, Categories } from '../pages';
import { appActions as actions } from '../redux/actions';
import { CONSTANTS } from '../utils/constants';
import { IApp } from '../interfaces';

// Defining Own Props to avoid unwanted props
interface IOwnProps {
  classes?: any;
}
type IProps = IOwnProps &
  IApp.StateToProps & 
  IApp.DispatchFromProps;

// Defining Own State to avoid unwanted props  
interface IState {}

// Mapping Store data with props 
const mapStateToProps = function(state: any){
  return {
    isDrawerOpen: state.app.isDrawerOpen,
  }
};

// Main Component
class AppMain extends React.Component<IProps, IState> {
 
  // Function to render UI
  render() {
    const { classes } = this.props;
    return (
        <Box className={clsx(classes.content, {
            [classes.contentShift]: this.props.isDrawerOpen,
            })} position={'relative'}>
            <div className={classes.drawerHeader} />
            <AppLoading />
            <Route exact={true} path="/" component={ListProduct}/>
            <Route path="/categories" component={Categories}/>
        </Box>
    );
  }
}
// High Order Component to connect with store and action mapping
const AppMainWrapper = connect(mapStateToProps, actions)(AppMain);

// styles
const styles = {
  content: {
    flexGrow: 1,
    marginLeft: -CONSTANTS.DRAWER_WIDTH,
    paddingTop: 80,
    paddingLeft: 25,
    paddingRight: 25,
    paddingBottom: 25, 
  },
  contentShift: {
    marginLeft: 0,
  }
}

// Defining proptery to Higher Order Component
AppMainWrapper.propTypes = {
  classes: PropTypes.object.isRequired,
}

/* High Order Component to apply styles as styles can either be applied in HOOK (not in React Component) 
  or with Higher Order Component. We used higher order component as we have used material components.
*/
const AppMainWithStyles = withStyles(styles)(AppMainWrapper)

export { AppMainWithStyles as AppMain }