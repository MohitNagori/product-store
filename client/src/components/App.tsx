import * as React from 'react';
import { BrowserRouter as Router } from "react-router-dom";
import { makeStyles, useTheme, Theme, createStyles, CssBaseline, }  from '@material-ui/core';
import { AppDrawer } from './AppDrawer';
import { AppHeader } from './AppHeader';
import { AppMain } from './AppMain';

// styles 
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
    },
  })
);

// App Component
export default function App() {
  const classes = useStyles();
  const theme = useTheme();

  return (
    <Router>
      <div className={classes.root}>
        <CssBaseline />
        <AppHeader/>
        <AppDrawer theme={theme}/>
        <AppMain/>  
      </div>
    </Router>
  );
}