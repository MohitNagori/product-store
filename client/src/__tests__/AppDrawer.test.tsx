import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { AppDrawer } from '../components';

describe('Testing AppDrawer component', () => {
  it('renders as expected', () => {
    const wrapper = shallow(<AppDrawer />);
    expect(toJson(wrapper)).toMatchSnapshot();
 });
});