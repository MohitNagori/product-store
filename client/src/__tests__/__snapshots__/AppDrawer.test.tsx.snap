// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AppDrawer component renders as expected 1`] = `
<Connect(AppDrawer)
  classes={
    Object {
      "drawer": "Connect(AppDrawer)-drawer-1",
      "drawerHeader": "Connect(AppDrawer)-drawerHeader-3",
      "drawerPaper": "Connect(AppDrawer)-drawerPaper-2",
      "link": "Connect(AppDrawer)-link-5",
      "menuItem": "Connect(AppDrawer)-menuItem-4",
    }
  }
/>
`;
